File: sections/06_two_pointers_sliding_windows/problems/a_sorted_two_sum/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        long long target;
        cin >> n >> target;
        vector<long long> values(n);
        for (long long &value : values) cin >> value;

        int left = 0;
        int right = n - 1;
        bool ok = false;
        while (left < right) {
            long long total = values[left] + values[right];
            if (total == target) {
                ok = true;
                break;
            }
            if (total < target) {
                ++left;
            } else {
                --right;
            }
        }
        cout << (ok ? "YES" : "NO") << '\n';
    }

    return 0;
}

File: sections/06_two_pointers_sliding_windows/problems/a_sorted_two_sum/solution.py
"""Reference solution for A. Sorted Two Sum."""

import sys


def exists_pair(values: list[int], target: int) -> bool:
    left = 0
    right = len(values) - 1
    while left < right:
        total = values[left] + values[right]
        if total == target:
            return True
        if total < target:
            left += 1
        else:
            right -= 1
    return False


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        target = data[idx + 1]
        idx += 2
        values = data[idx:idx + n]
        idx += n
        out.append("YES" if exists_pair(values, target) else "NO")
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

